refresh.template.py

```diff
@@ -90,6 +97,57 @@ def _get_headers(compile_args: typing.List[str], source_path_for_sanity_check: t
     return headers
 
 
+def _get_headers_msvc(compile_args: typing.List[str]):
+    # We don't want to compile, but only pre-process and show all the included header files.
+    # We don't want to have the preprocessed results written to a file, so we get them printed
+    # to `stderr`.
+    header_cmd = list(compile_args) + ["/showIncludes", "/EP"]
+
+    # cl.exe expects `INCLUDE` (and normally `LIB` but that is irrelevant here) to be set so that it can find the
+    # system headers. We have no idea what the correct paths are. So in case these variables are not set, we need
+    #  to infer them somehow. These variables might not be set if the user has a custom `cc_toolchain()`
+    # configured with `bazel` and does not rely on having sourced vcvarsall.bat.
+    environ = dict(os.environ)
+    if "INCLUDE" not in environ:
+        # We assume the base MSVC install based based on the `cl.exe` path in our command.
+        base = pathlib.Path(header_cmd[0]).parent.parent.parent.parent
+        # And we find the newest Windows 10 kits version that exists.
+        windows_kits = sorted( pathlib.Path("C:\\Program Files (x86)\\Windows Kits\\10\\include").glob("*"))[-1]
+        environ["INCLUDE"] =  os.pathsep.join(
+            (
+                os.fspath(base / "ATLMFC/include"),
+                os.fspath(base / "include"),
+                os.fspath(windows_kits / "ucrt"),
+                os.fspath(windows_kits / "shared"),
+                os.fspath(windows_kits / "um"),
+                os.fspath(windows_kits / "winrt"),
+                os.fspath(windows_kits / "cppwinrt"),
+            )
+        )
+
+    header_search_process  = subprocess.run(
+        header_cmd,
+        cwd=os.environ["BUILD_WORKSPACE_DIRECTORY"],
+        stderr=subprocess.PIPE,
+        stdout=subprocess.DEVNULL,
+        env=environ,
+        encoding="utf-8",
+        check=False, # We explicitly ignore errors and carry on.
+    )
+
+    if not header_search_process.stderr:
+        return [] # Worst case, we couldn't get the headers,
+
+    headers = (
+        # TODO This is not enough. Based on the locale, `cl.exe` will emit different marker strings. This should be extendend to multiple languages.
+        line.replace("Note: including file:", "").strip()
+        for line in header_search_process.stderr.splitlines()
+        if "Note: including file:" in line
+    )
+
```

Comment:
Worth also printing any real errors?

Reply:
I am not sure how to detect real errors. Because `cl.exe` prints the pre-processed file to `stderr`. But also some error-lines. Will look into it.

Reply:
Done.

Reply:
How about just printing all the lines that don't match as starting with "Note: including file:"?

...at least if there's not too much other logging? I'd just like to prevent silent failures for people, especially given msvc being hard to parse. Totally get that it's annoying to parse both out of stderr.

Also, maybe `startswith`?

Reply:
I think I saw some stuff during development, but I am unable to create a reproducer that produces a lot of spam. So let's try this for now and see how that behaves.

Reply:
Awesome. Thanks for your care and double checking.